pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,844 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	selectorSyncSets, err := r.getSelectorSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Apply SyncSets
+	syncStatusesForSyncSets, syncSetsNeedRequeue := r.applySyncSets(
+		"SyncSet",
+		syncSets,
+		clusterSync.Status.SyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SyncSets = syncStatusesForSyncSets
+
+	// Apply SelectorSyncSets
+	syncStatusesForSelectorSyncSets, selectorSyncSetsNeedRequeue := r.applySyncSets(
+		"SelectorSyncSet",
+		selectorSyncSets,
+		clusterSync.Status.SelectorSyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SelectorSyncSets = syncStatusesForSelectorSyncSets
+
+	setFailedCondition(clusterSync, needToDoFullReapply)
+
+	// Create or Update the ClusterSync
+	if needToCreateClusterSync {
+		logger.Info("creating ClusterSync")
+		clusterSync.Namespace = cd.Namespace
+		clusterSync.Name = cd.Name
+		clusterSync.OwnerReferences = []metav1.OwnerReference{{
+			APIVersion:         hivev1.SchemeGroupVersion.String(),
+			Kind:               "ClusterDeployment",
+			Name:               cd.Name,
+			UID:                cd.UID,
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}}
+		if err := r.Create(context.Background(), clusterSync); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create ClusterSync")
+			return reconcile.Result{}, err
+		}
+	} else if !reflect.DeepEqual(origStatus, &clusterSync.Status) {
+		logger.Info("updating ClusterSync")
+		if err := r.Status().Update(context.Background(), clusterSync); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update ClusterSync")
+			return reconcile.Result{}, err
+		}
+	}
+
+	if needToDoFullReapply {
+		logger.Info("setting last full apply time")
+		lease.Spec.RenewTime = metav1.NowMicro()
+		if needToCreateLease {
+			logger.Info("creating lease for ClusterSync")
+			lease.Namespace = leaseKey.Namespace
+			lease.Name = leaseKey.Name
+			lease.OwnerReferences = []metav1.OwnerReference{{
+				APIVersion:         hivev1.SchemeGroupVersion.String(),
+				Kind:               "ClusterSync",
+				Name:               clusterSync.Name,
+				UID:                clusterSync.UID,
+				BlockOwnerDeletion: pointer.BoolPtr(true),
+			}}
+			if err := r.Create(context.Background(), lease); err != nil {
+				logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create lease for ClusterSync")
+				return reconcile.Result{}, err
+			}
+		} else {
+			logger.Info("updating lease for ClusterSync")
+			if err := r.Update(context.Background(), lease); err != nil {
+				logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update lease for ClusterSync")
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	result := reconcile.Result{Requeue: true, RequeueAfter: r.timeUntilFullReapply(lease)}
```

Comment:
Hm. Given we're jittery, this will almost never be right.

If the jitter is smaller here than on L302 during the subsequent reconcile, we'll skip updating everything until the *next* reconcile, thus roughly doubling the max expected full-reapply time.

If the opposite, we're effectively ignoring the jitter on L302.

This may be slightly mitigated if there are [s]ss or cd updates to trigger a reconcile early. But steady state is pretty common, yah?

Seems like the right way to do this is to calculate `timeUntilFullReapply` once, and store *that* stamp in the `ClusterLease` (heck, you don't even need to rename the field). Then the calculation on L302 is simply whether `Now()` is later than that.

Reply:
I do not follow along with how we would be doubling the max expected full-reapply time. Could you give me an example with concrete numbers?

Here is my example.
L302: time unitl full reapply = 30 minutes + 10 minute jitter = 40 minutes
  - no re-apply
L391: time until full reapply = 30 minutes + 5 minutes jitter = 35 minutes
  - wait 35 minutes before next reconcile
L302: time until full reapply = -5 minutes + 10 minute jitter = 5 minutes
  - no re-apply
L391: time until full reapply = -5 minutes + 2 minutes jitter = 0 minutes (forced to 0)
  - wait a small delay before next reconcile
L302: time until full reapply = -5 minutes + 10 minute jitter = 5 minutes
  - no re-apply
L391: time until full reapply = -5 minutes + 2 minutes jitter = 0 minutes (forced to 0)
  - wait another slightly longer delay before next reconcile
Eventually, either L302 will get a time until full reapply that is 0 either because the random jitter is small enough or because the time since the last re-apply is greater than 132 minutes. The worst case that I can see is waiting until (120 minutes + 1000 seconds) = 136 minutes, 40 seconds since the last re-apply.

Reply:
Man, I was really asleep at the wheel last night, wasn't I?

You're right, the doubling isn't a thing. We'll end up averaging higher than the halfway point of the jitter range, but that's no big deal.

Sorry for the noise.

Reply:
BTW, I should note that I was surprised that the jitter is always *added*. I initially thought/assumed the idea was for the *average* to be 2h, with $jitter to either side.